src/dashboardGenerator.js

```diff
@@ -0,0 +1,359 @@
+                  "value": 7
+                },
+                {
+                  "value": 1,
+                  "label": "1 day"
+                }
+              ],
+              {
+                "color": ANNOTATION_MEDIUM_COLOUR,
+                "label": "< 1 day",
+                "value": 1,
+                "fill": "below"
+              }
+            ]
+          }
+        }
+      }
+    ];
+
+    return state;
+  }
+
+  getMetrics(state) {
+    return new Promise(function (resolve, reject) {
+      state.cloudwatch.listMetrics({ "Namespace": "Operations" }).eachPage(function (err, data) {
+        if (err) {
+          reject(err);
+          return;
+        }
+
+        if (data === null) {
+          resolve(state);
+        } else {
+
+          state.metrics =
+            data.Metrics.map(m => m.Dimensions.filter(d => d.Name === 'service').map(d => { return { serviceName: d.Value, metricName: m.MetricName } }))
+              .reduce((a, b) => a.concat(b), state.metrics);
+        }
+      });
+
+    });
+  }
+
+
+  putDashboard(state) {
+    state.metrics = [...new Set(state.metrics)].sort();
+    let y = 0; // leave space for the legend on first row
+
+    applyLimits(state);
+
+    let dashboard = {
+      "start": "-P42D",
+      "widgets": [],
+    };
+
+    const TEXT_HEIGHT = 4;
+    let x = 0;
+    [
+      {
+        "title": "MTTR",
+        "description":
+          "Mean time to recovery.  **Lower = better**\n\n" +
+          "When a failure occurs, how quickly is service restored.\n\n" +
+          "When the system fails, the team should \"stop the line\" and swarm to fix it.\n"
+      },
+      {
+        "title": "MTBF",
+        "description":
+          "Mean time between failures.  **Higher = better**\n\n" +
+          "MTBF describes system stability.  How often does the service fail.\n\n" +
+          "An unstable system suggests systemic quality/resiliency issues that needs to be addressed.\n"
+      },
+      {
+        "title": "MTTF",
+        "description":
+          "Mean time to failure.  **Higher = better**\n\n" +
+          "How stable/available is the system.\n\n" +
+          "A low MTTF indicates that the system if often unavailable.\n\n" +
+          "Improve availability by improving system resiliency.  Reduce planned outages using staged deployment, such as blue/green or canary releasing.\n"
+      }
+    ].forEach(l => {
+      dashboard.widgets.push({
+        "type": "text",
+        "x": x,
+        "y": y,
+        "width": l.y ? l.y : 8,
+        "height": TEXT_HEIGHT,
+        "properties": {
+          "markdown": `\n### ${l.title}\n${l.description}`
+        }
+      });
+
+      x += 8;
+    });
+    y += TEXT_HEIGHT;
+
+    let widget = [trendWidgets(state.metrics, y, state)].concat(averageWidgets(state.metrics, y, state));
+    y += WIDGET_HEIGHT;
+
+    // flatten the nested arrays
+    dashboard.widgets = [].concat.apply(dashboard.widgets, widget);
```

Comment:
Where's the nesting occur?

can you just use `dashboard.widgets.push` rather than 
`let widget = [trendWidgets(state.metrics, y, state)].concat(averageWidgets(state.metrics, y, state));`?

Reply:
Push doesn't work

Reply:
Done